src/game/AI/ScriptDevAI/scripts/kalimdor/dustwallow_marsh.cpp

```diff
@@ -1004,6 +1007,192 @@ UnitAI* GetAI_mob_invis_firework_helper(Creature* pCreature)
     return new mob_invis_firework_helper(pCreature);
 }
 
+/*######
+## npc_theramore_practicing_guard
+######*/
+enum
+{
+    NPC_THERAMORE_COMBAT_DUMMY = 4952,
+};
+
+struct npc_theramore_practicing_guardAI : public ScriptedAI
+{
+    npc_theramore_practicing_guardAI(Creature* pCreature) : ScriptedAI(pCreature)
+    {
+        Reset();
+        GetNearbyDummyIfNotExist();
+    }
+
+    uint32 m_attackTimer;
+    uint32 m_breakTimer;
+    uint32 m_sitTimer;
+    bool m_bisAttacking;
+    bool m_bsitDown, m_bstandUp;
+    bool m_binCombatWithPlayer;
+    ObjectGuid attackableDummy;
+
+    void DoCallForHelp(float) override {}
+
+    void HandleAssistanceCall(Unit*, Unit*) override {}
+
+    void GetNearbyDummyIfNotExist()
+    {
+        if (!attackableDummy)
+        {
+            if (Unit* nearbyDummy = GetClosestCreatureWithEntry(m_creature, NPC_THERAMORE_COMBAT_DUMMY, 2.f))
+            {
+                attackableDummy = nearbyDummy->GetObjectGuid();
+            }
+        }
+    }
+
+    void EnterEvadeMode() override
+    {
+        ScriptedAI::EnterEvadeMode();
+        m_binCombatWithPlayer = false;
+    }
+
+    void JustReachedHome() override
+    {
+        Reset();
+        SetRootSelf(true);
+        GetNearbyDummyIfNotExist();
+    }
+
+    void EnterCombat(Unit* who) override
+    {
+        if (who->GetEntry() == NPC_THERAMORE_COMBAT_DUMMY)
+            return;
+
+        if (who->GetObjectGuid() != attackableDummy)
+        {
+            m_binCombatWithPlayer = true;
+            SetRootSelf(false);
+            ScriptedAI::EnterCombat(who);
+        }
+    }
+
+    void AttackedBy(Unit* who) override
+    {
+        if (who->GetObjectGuid() != attackableDummy)
+        {
+            m_binCombatWithPlayer = true;
+            SetRootSelf(false);
+            ScriptedAI::AttackedBy(who);
+        }
+    }
+
+    void Reset() override
+    {
+        m_attackTimer = 130 * IN_MILLISECONDS;
+        m_breakTimer = 20 * IN_MILLISECONDS;
+        m_bisAttacking = false;
+        m_bsitDown = false;
+        m_bstandUp = false;
+        m_binCombatWithPlayer = false;
+        m_creature->SetStandState(UNIT_STAND_STATE_STAND);
+    }
+
+    void UpdateAI(const uint32 diff) override
+    {
+        if (!attackableDummy && !m_creature->IsInCombat() && !m_creature->IsMoving())
```

Comment:
Consider adding a timer here so this is not called every app heartbeat.

Reply:
it's called every heartbeat, but most heartbeats shouldn't have an uninitialized `attackableDummy` ObjectGuid. I think that has less impact than actually checking for a timer.